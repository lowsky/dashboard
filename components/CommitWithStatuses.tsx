import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
    faCheckCircle,
    faHourglass,
    faExclamationCircle,
    faTimes,
}  from '@fortawesome/free-solid-svg-icons'

import { GithubCommit } from '../lib/types/resolvers';

import './CommitWithStatuses.module.css';

function icon4context(context = '', avatar_url: string) {
    if (avatar_url) {
        return <img className="contextlogo" width={24} height={24} src={avatar_url} alt={context} />;
    }
    return <span>{context}</span>;
}

function icon4status(status: 'success' | 'pending' | 'failure' | 'error' | any) {
    const style = {
        verticalAlign: 'top'
    };
    if (status === 'success') {
        return <FontAwesomeIcon style={style} icon={faCheckCircle} />;
    }
    if (status === 'pending') {
        return <FontAwesomeIcon style={style} icon={faHourglass} />
    }
    if (status === 'failure') {
        return <FontAwesomeIcon style={style} icon={faExclamationCircle } />
    }
    if (status === 'error') {
        return <FontAwesomeIcon style={style} icon={faTimes } />
    }
    return <span>{status}</span>;
}

function status2color(status) {
    if (status === 'success') {
        return 'green';
    }
    if (status === 'pending') {
        return 'orange';
    }
    if (status === 'failure') {
        return 'red';
    }
    if (status === 'error') {
        return 'red';
    }
    return 'inherit';
}

const renderStatus = ({ target_url, avatar_url, context, description, state }, idx) => (
    <a key={idx}
        className="commitLink"
        href={target_url}
        style={{ color: status2color(state) }}
        title={context + ' ' + description}>
            {icon4context(context, avatar_url)}
            {icon4status(state)}
        {' '}
    </a>
);

export interface CommitWithStatusProps {
    commit?: GithubCommit;
}
let CommitWithStatus: React.FC<CommitWithStatusProps> = ({ commit = {} }) => {
    const { sha = '<missing>', date = '', message = '<missing>', status = [],
      } = commit;
    const author = {
        login: '',
        name: '',
        email: '',
        avatar_url: '',
        ...commit.author
    };
    const githubCommit = `https://github.com/lowsky/dashboard/tree/${sha}`;

    const onlyTakeFirstStatusPerContext = (statuses) => {
        const foundContexts = [];

        return statuses.reduce((acc, item) => {
            if (foundContexts[item.context]) {
                return acc;
            }
            foundContexts[item.context] = item;
            return [...acc, item];
        }, []);
    };

    return (
        <>
            <div>
                <a href={githubCommit}>
                    <b>{message.split('\n\n', 1)}</b>
                </a>
            </div>
            <div>
                <i>{date}</i>
                &nbsp; by &nbsp;
                <a href={`https://github.com/${author.login}`}>
                    {author.avatar_url && (
                        <img className="commit_avatar" width={32} src={author.avatar_url} alt="avatar" />
                    )}
                    &nbsp;
                    <span>{author.login}</span>
                </a>
                &nbsp;
                {author.email && <a href={'mailto:' + author.email}>{'?' + author.name}</a>}
            </div>
            <div className="statusline">
                {onlyTakeFirstStatusPerContext(status).map((s, idx) => renderStatus(s, idx))}
            </div>
        </>
    );
};

export default CommitWithStatus;
