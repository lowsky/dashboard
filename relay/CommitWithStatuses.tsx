import { createFragmentContainer, graphql } from 'react-relay';

import CommitWithStatus from '../components/CommitWithStatuses';

export const fakeCommitWithStatuses = {
    sha: '',
    message: 'fake commit.',
    date: '01.01.2016',
    status: [
        {
            context: 'check',
            description: 'some pre-commit check',
            sate: 'unknown',
            target_url: 'http://precommit.check/',
            avatar_url:
                'https://avatars1.githubusercontent.com/oa/581325?s=40&u=3d11470b9117a9024372113ecd1fc682b02395b8&v=4/',
            updated_at: new Date().toISOString(),
        },
    ],
};

export default createFragmentContainer(CommitWithStatus, {
    commit: graphql`
        fragment CommitWithStatuses_commit on GithubCommit {
            sha
            message
            date
            status {
                context
                description
                state
                target_url
                updated_at
                avatar_url
            }
        }
    `,
});
