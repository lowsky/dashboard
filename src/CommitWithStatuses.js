import React from 'react';
import Relay from 'react-relay';

function icon4context(context) {
    if(context=== 'bitHound - Code') {
        return (<img width={48} height={48} src="./bithound.png" alt="bitHound - Code"/>);
    }
    if(context=== 'bitHound - Dependencies') {
        return (<img width={48} height={48} src="./bithound.png" alt="bitHound - Dependencies"/>);
    }
    if(context=== 'codacy/pr') {
        return (<img width={48} height={48} src="./codacy.png" alt="codacy"/>);
    }
    if(context=== 'ci/circleci') {
        return (<img width={48} height={48} src="./circleci.png" alt="circleci"/>);
    }
    if(context=== 'code-review/gitcolony') {
        return (<img width={48} height={48} src="./bithound.png" alt="gitcolony"/>);
    }

    return (<span>{ context }</span> );
}

function icon4status(status) {
    if(status=== 'success') {
        return (<span className="glyphicon glyphicon-ok green"></span>);
    }
    if(status=== 'pending') {
        return (<span className="glyphicon glyphicon-hourglass orange"></span>);
    }
    if(status=== 'failure') {
        return (<span className="glyphicon glyphicon-remove red"></span>);
    }
    if(status=== 'error') {
        return (<span className="glyphicon glyphicon-remove red"></span>);
    }
    return (<span>{ status }</span> );
}

function status2color(status) {
    if(status=== 'success') {
        return "green";
    }
    if(status=== 'pending') {
        return "orange";
    }
    if(status=== 'failure') {
        return "red";
    }
    if(status=== 'error') {
        return "red";
    }
    return 'inherit';
}

function renderStatus(status) {
    const linkStyle = {};
    const color = status2color(status.state);

    return (
        <div><a className='commitLink' href={ status.target_url } style={ {color: color} }>
            { icon4status(status.state) } - { icon4context(status.context) } { status.description }</a>
        </div>
        );
}

let LastCommit = React.createClass({
    render: function () {
        const { sha='<missing>', date='', message='<missing>', status=[] } = this.props.commit;

        const githubLastCommit = `https://github.com/lowsky/dashboard/tree/${sha}`;

        return (
            <div>
            <a href={githubLastCommit}>
                <span><i>{ date }</i></span>
                <b>{ message }</b>
            </a>
                {
                    status.map(s => renderStatus(s))
                }
            </div>
        );
    }
});

export default LastCommit;
