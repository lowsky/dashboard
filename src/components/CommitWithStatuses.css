.level {
    justify-content: start;
}

.contextlogo {
    verticalAlign: top;
}

.statusicon {
    vertical-align: top;
    margin-left: -0.5em;
    margin-top: -0.5em;
    padding-right: 0.25em;
    background: white;
    z-index: 4;
}
