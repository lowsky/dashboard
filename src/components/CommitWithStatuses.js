import React from 'react';
import PropTypes from 'prop-types';

import bithoundLogo from './bithound.png';
import circleciLogo from './circleci.svg';
import codacyLogo from './codacy.png';
import snykLogo from './snyk.svg';

const ContextLogo = props => <img width={48} height={48} {...props} style={{ verticalAlign: 'top' }} alt="" />;

function icon4context(context) {
    if (context === 'bitHound - Code') {
        return <ContextLogo src={bithoundLogo} alt="bitHound - Code" />;
    }
    if (context === 'bitHound - Dependencies') {
        return <ContextLogo src={bithoundLogo} alt="bitHound - Dependencies" />;
    }
    if (context === 'codacy/pr') {
        return <ContextLogo src={codacyLogo} alt="codacy" />;
    }
    if (context === 'ci/circleci') {
        return <ContextLogo src={circleciLogo} alt="circleci" />;
    }

    if (context === 'security/snyk') {
        return <ContextLogo src={snykLogo} alt="snyk" />;
    }

    return (
        <span>
            {context}
        </span>
    );
}

function icon4status(status) {
    const StatusIcon = ({ type }) =>
        <i className="material-icons" style={{ fontSize: '48px' }}>
            {type}
        </i>;
    StatusIcon.propTypes = {
        type: PropTypes.string,
    };

    if (status === 'success') {
        return <StatusIcon type="check" />;
    }
    if (status === 'pending') {
        return <StatusIcon type="hourglass_empty" />;
    }
    if (status === 'failure') {
        return <StatusIcon type="cancel" />;
    }
    if (status === 'error') {
        return <StatusIcon type="error" />;
    }
    return (
        <span>
            ?{status}{' '}
        </span>
    );
}

function status2color(status) {
    if (status === 'success') {
        return 'green';
    }
    if (status === 'pending') {
        return 'orange';
    }
    if (status === 'failure') {
        return 'red';
    }
    if (status === 'error') {
        return 'red';
    }
    return 'inherit';
}

function renderStatus(status, idx) {
    return (
        <div key={idx} style={{ display: 'inline' }}>
            <a
                className="commitLink"
                href={status.target_url}
                style={{ color: status2color(status.state) }}
                title={status.context + ' ' + status.description}>
                {icon4context(status.context)}
                {icon4status(status.state)}
            </a>
        </div>
    );
}

let CommitWithStatus = ({ commit = {} }) => {
    const { sha = '<missing>', date = '', message = '<missing>', status = [] } = commit;
    const { author = {} } = commit;

    const githubCommit = `https://github.com/lowsky/dashboard/tree/${sha}`;

    const onlyTakeFirstStatusPerContext = statuses => {
        const foundContexts = [];

        return statuses.reduce((acc, item) => {
            if (foundContexts[item.context]) {
                return acc;
            }
            foundContexts[item.context] = item;
            return [...acc, item];
        }, []);
    };

    return (
        <div key={sha}>
            <a href={githubCommit}>
                <b>
                    {message.split('\n\n', 1)}
                </b>
            </a>
            <br />
            <span>
                <i>
                    {date}
                </i>
            </span>{' '}
            &nbsp; by &nbsp;
            <img width={32} src={author.avatar_url} alt="author-avatar" /> &nbsp;
            {author.login}
            <a href={'mailto:' + author.email}>{author.name}</a>
            <br />
            {onlyTakeFirstStatusPerContext(status).map((s, idx) => renderStatus(s, idx))}
        </div>
    );
};

CommitWithStatus.propTypes = {
    commit: PropTypes.object,
};

export default CommitWithStatus;
