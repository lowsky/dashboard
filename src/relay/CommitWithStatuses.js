import { createFragmentContainer, graphql } from 'react-relay';

import CommitWithStatus from '../components/CommitWithStatuses';

export const fakeCommitWithStatuses = {
    sha: '',
    message: 'fake commit.',
    date: '01.01.2016',
    status: [
        {
            context: 'check',
            description: 'some pre-commit check',
            sate: 'unknown',
            target_url: 'http://precommit.check/',
            updated_at: new Date().toISOString(),
        },
    ],
};

export default createFragmentContainer(
    CommitWithStatus,
    graphql`
        fragment CommitWithStatuses_commit on GithubCommit {
            sha
            message
            date
            status {
                context
                description
                state
                target_url
                updated_at
            }
        }
    `
);
