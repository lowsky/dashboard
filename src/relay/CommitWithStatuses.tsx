import { createFragmentContainer } from 'react-relay';
import graphql from 'babel-plugin-relay/macro';

import CommitWithStatus from '../components/CommitWithStatuses';

export const fakeCommitWithStatuses = {
    sha: '',
    message: 'fake commit.',
    date: '01.01.2016',
    status: [
        {
            context: 'check',
            description: 'some pre-commit check',
            sate: 'unknown',
            target_url: 'http://precommit.check/',
            updated_at: new Date().toISOString(),
        },
    ],
};

export default createFragmentContainer(CommitWithStatus, {
    commit: graphql`
        fragment CommitWithStatuses_commit on GithubCommit {
            sha
            message
            date
            status {
                context
                description
                state
                target_url
                updated_at
            }
        }
    `,
});
