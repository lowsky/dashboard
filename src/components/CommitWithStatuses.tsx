import React from 'react';
import PropTypes from 'prop-types';

import bithoundLogo from './bithound.png';
import circleciLogo from './circleci.svg';
import codacyLogo from './codacy.png';
import snykLogo from './snyk.svg';

import './CommitWithStatuses.css';

const ContextLogo = props => <img className="contextlogo" width={24} height={24} {...props} alt="" />;

function icon4context(context = '') {
    if (context === 'bitHound - Code') {
        return <ContextLogo src={bithoundLogo} alt="bitHound - Code" />;
    }
    if (context === 'bitHound - Dependencies') {
        return <ContextLogo src={bithoundLogo} alt="bitHound - Dependencies" />;
    }
    if (context.indexOf('Codacy/PR') >= 0) {
        return <ContextLogo src={codacyLogo} alt="codacy" />;
    }
    if (context.indexOf('ci/circleci') >= 0) {
        return <ContextLogo src={circleciLogo} alt="circleci" />;
    }

    if (context.indexOf('security/snyk') >= 0) {
        return <ContextLogo src={snykLogo} alt="snyk" />;
    }

    return <span>{context}</span>;
}

function icon4status(status: string) {
    const StatusIcon: React.FC<{ type: string }> = ({ type }) => <i className={`fa fa-${type} statusicon`} />;

    StatusIcon.propTypes = {
        type: PropTypes.string,
    };

    if (status === 'success') {
        return <StatusIcon type="check-circle" />;
    }
    if (status === 'pending') {
        return <StatusIcon type="hourglass" />;
    }
    if (status === 'failure') {
        return <StatusIcon type="exclamation-circle" />;
    }
    if (status === 'error') {
        return <StatusIcon type="times" />;
    }
    return <span>?{status} </span>;
}

function status2color(status: string) {
    if (status === 'success') {
        return 'green';
    }
    if (status === 'pending') {
        return 'orange';
    }
    if (status === 'failure') {
        return 'red';
    }
    if (status === 'error') {
        return 'red';
    }
    return 'inherit';
}

const renderStatus = ({ target_url, context, description, state }, idx) => (
    <span key={idx}>
        <a
            className="commitLink"
            href={target_url}
            style={{ color: status2color(state) }}
            title={context + ' ' + description}>
            {icon4context(context)}
            {icon4status(state)}
        </a>
    </span>
);

let CommitWithStatus = ({ commit = {} }) => {
    // @ts-ignore
    const { sha = '<missing>', date = '', message = '<missing>', status = [], author = {} } = commit;

    const githubCommit = `https://github.com/lowsky/dashboard/tree/${sha}`;

    const onlyTakeFirstStatusPerContext = statuses => {
        const foundContexts = [];

        return statuses.reduce((acc, item) => {
            if (foundContexts[item.context]) {
                return acc;
            }
            foundContexts[item.context] = item;
            return [...acc, item];
        }, []);
    };

    return (
        <>
            <div>
                <a href={githubCommit}>
                    <b>{message.split('\n\n', 1)}</b>
                </a>
            </div>
            <div>
                <i>{date}</i>
                &nbsp; by &nbsp;
                <a href={`https://github.com/${author.login}`}>
                    {author.avatar_url && (
                        <img className="commit_avatar" width={32} src={author.avatar_url} alt="avatar" />
                    )}
                    &nbsp;
                    <span>{author.login}</span>
                </a>
                &nbsp;
                {author.email && <a href={'mailto:' + author.email}>{'?' + author.name}</a>}
            </div>
            <div className="statusline">
                {onlyTakeFirstStatusPerContext(status).map((s, idx) => renderStatus(s, idx))}
            </div>
        </>
    );
};

export default CommitWithStatus;
