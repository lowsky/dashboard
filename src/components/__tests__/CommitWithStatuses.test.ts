import React from 'react';
import ReactDOM from 'react-dom';

import CommitWithStatuses from '../CommitWithStatuses';
import renderer from 'react-test-renderer';

export const status = {
    target_url: 'target_url',
    state: 'state',
    description: 'description',
    context: 'context',
};
export const commit = {
    sha: '<missing>',
    date: '01.01.2001',
    message: 'some-message,',
    status: [status],
    author: {
        avatar_url: 'avatar_url',
        name: 'name',
        login: 'login',
        email: 'email',
    },
};

describe('CommitWithStatuses component', () => {
    it('should render without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(<CommitWithStatuses />, div);
    });

    it('should render as expected', () => {
        const props = {
            commit,
        };
        const appContainer = renderer.create(<CommitWithStatuses {...props} />).toJSON();
        expect(appContainer).toMatchSnapshot();
    });
});
