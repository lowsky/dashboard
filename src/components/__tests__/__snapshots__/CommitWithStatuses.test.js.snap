// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommitWithStatuses component should render as expected 1`] = `
<div>
  <div>
    <a
      href="https://github.com/lowsky/dashboard/tree/<missing>"
    >
      <b>
        some-message,
      </b>
    </a>
  </div>
  <div
    className="level"
    style={
      Object {
        "justifyContent": "normal",
      }
    }
  >
    <i>
      01.01.2001
    </i>
      by  
    <img
      alt="avatar"
      src="avatar_url"
      width={32}
    />
      
    <span>
      login
    </span>
      
    <a
      href="mailto:email"
    >
      ?name
    </a>
  </div>
  <div
    style={
      Object {
        "display": "inline",
      }
    }
  >
    <a
      className="commitLink"
      href="target_url"
      style={
        Object {
          "color": "inherit",
        }
      }
      title="context description"
    >
      <span>
        context
      </span>
      <span>
        ?
        state
         
      </span>
    </a>
  </div>
</div>
`;
